### PR TITLE
Add network connection local support to work with collections

### PR DESCRIPTION
**Pull request by ganeshrn**
##### SUMMARY

*  The legacy behavior of network action plugins using persistent
   the framework is to override the connection=local with the actual connection type (network_cli/netconf/httpapi) based
   on the value of transport option in provider.
*  After the actual connection plugin is identified try to load it from
   `ansible.netcommon` collection, if it fails load it from ansible
   core. This is done to work with Ansible 2.10 and also maintain backward compatibility for
   Ansible 2.9 version to ensure it works with network collections.
*  To support this need to pass collection_list value to connection
   plugins. As in case of connection=local it loads `persistent`
   connection within action plugin which in turn invokes `network_cli`
   or `netconf` connection. Similarly `network_cli` connection plugin invokes
   `terminal` and `cliconf plugins and `netconf` connection plugin invokes `netconf`
   `plugins` based on value on `ansible_network_os`.
*  Add deprecation warning for connection=local support and
   provider support for applicable platforms

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

cli/scripts/ansible_connection_cli_stub.py
executor/task_executor.py
module_utils/network/eos/eos.py
module_utils/network/ios/ios.py
plugins/action/__init__.py
plugins/action/eos.py
plugins/action/ios.py
plugins/cliconf/__init__.py
plugins/connection/httpapi.py
plugins/connection/netconf.py
plugins/connection/network_cli.py
plugins/connection/persistent.py
plugins/httpapi/__init__.py
plugins/loader.py
plugins/netconf/__init__.py
plugins/terminal/__init__.py
##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: lib/ansible/plugins/action/aireos.py [hunk 1]
```diff
@@ -38,42 +38,62 @@ def run(self, tmp=None, task_vars=None):
...
+
+        if persistent_connection == 'network_cli':
+            provider = self._task.args.get('provider', {})
+            if any(provider.values()):
+                display.warning('provider is unnecessary when using network_cli and will be ignored')
+                del self._task.args['provider']
+        elif self._play_context.connection == 'local':
+            provider = load_provider(aireos_provider_spec, self._task.args)
+            pc = copy.deepcopy(self._play_context)
+            pc.connection = 'ansible.netcommon.network_cli'
```
Comment (pabelanger) [inline lib/ansible/plugins/action/aireos.py:52]: How can we hardcode this here? This means that ansible.netcommon namespace is now locked into our codebase.

File: lib/ansible/plugins/action/aireos.py [hunk 2]
```diff
@@ -38,42 +38,62 @@ def run(self, tmp=None, task_vars=None):
...
+            pc.network_os = 'aireos'
+            pc.remote_addr = provider['host'] or self._play_context.remote_addr
+            pc.port = int(provider['port'] or self._play_context.port or 22)
+            pc.remote_user = provider['username'] or self._play_context.connection_user
+            pc.password = provider['password'] or self._play_context.password
+            command_timeout = int(provider['timeout'] or C.PERSISTENT_COMMAND_TIMEOUT)
+
+            connection = self._shared_loader_obj.connection_loader.get('ansible.netcommon.persistent', pc, sys.stdin,
```
Comment (pabelanger) [inline lib/ansible/plugins/action/aireos.py:60]: same comment

File: lib/ansible/plugins/action/aireos.py (hunk 1 above)
Comment (ganeshrn) [reply lib/ansible/plugins/action/aireos.py:52]: Yes, the `ansible.netcommon` is locked and will be used to distribute all the common network code.

**Review by pabelanger — APPROVED** [review]
I didn't get a chance to test this will our collections, I think we can rely on your local tests for now.  Once this merges, we'll sync into collections and then able to confirm. If issues, we can continue to iterate on it.

Thoughts?

**Comment by ganeshrn** [comment]
@pabelanger Sure. Works for me.

**Comment by pabelanger** [comment]
recheck

**Comment by ganeshrn** [comment]
recheck

**Comment by nitzmahone** [comment]
This looks like a misuse/misunderstanding of what the collection list is for. Either ansible_network_os should be specified fully-qualified (the way it should be for new collection-native stuff), or for backwards compatibility with existing unqualified values, you can either have a hardcoded internal translation to the collection-qualified version when looking it up, or once we have all the tombstoning/redirection stuff in place in the plugin loader, you should be able to use that. Plumbing the collection list through to ansible-connection shouldn't be necessary, and will likely cause other problems.

**Comment by ganeshrn** [comment]
@nitzmahone Thanks for looking into it.

For connection=local with network legacy use case, the network os value is hardcoded in action plugin eg: https://github.com/ansible/ansible/blob/a92109363f1a454d368625dd8e2a0c0f3b74298a/lib/ansible/plugins/action/ios.py#L52

As you mentioned we can hardcode the fully-qualified network os name in action plugin itself which wouldn't require passing around the collection list. However, afaik we have finalised the collection namespace for 6 platforms that we support. For remaining platforms, it is still unknown.
If the community collection namespace is finalised which will have code for community-supported network platforms I can hardcode that fully qualified network os name in the individual action plugins.

**Comment by ganeshrn** [comment]
@nitzmahone @pabelanger I have split the changes in orginal PR into PR's
1) PR https://github.com/ansible/ansible/pull/66302 to handle network platforms for which collection namespace is already identified.

2) This PR can be used to track the changes in case of network platforms for which collection namespace is not yet identified. Currently the `ansible.community` namespace prefix for network os value is added as a placeholder in action plugins and it will be changed when the final namespace is identified.

**Comment by ganeshrn** [comment]
Updated the PR to reflect the values of network_os in action plugin based on the scenario files here https://github.com/ansible-community/collection_migration/tree/master/scenarios/nwo
The PR is ready for review and merge
ready_for_review

**Comment by justjais** [comment]
recheck

**Review by justjais — APPROVED** [review]

**Comment by ganeshrn** [comment]
Moved to PR https://github.com/ansible-collections/community.network/pull/120

**Pull request closed without merging**